Bluetooth: controller: Add connectable extended advertising

andrzej-kaczmarek opened the pull request:
This adds support for connectable extended advertising instance.

- [x] process AUX_CONNECT_REQ and send back AUX_CONNECT_RSP
- [x] establish connection
- [x] stop aux ticker on connection created
- [x] free resources if AUX_CONNECT_RSP failed to be sent

update:
babblesim testcases will not be added here since we do not support connecting to extended instances thus such testcase will always fail

andrzej-kaczmarek committed:
Bluetooth: controller: Add public helper to process connection ind

We will use the same code for auxes so let's make it public.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Use decoded AdvA/TargetA for conn ind helper

The same helper will be used for legacy and extended advertising so we
better pass AdvA and TargetA directly instead of complete advertising
PDU. This is because those fields are at different locations in legacy
and extended advertising PDUs so caller can figure them out.

Also, we can now check for directed advertising using TargetA since it
will be only supplied for such advertising (NULL otherwise).

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Use proper transmitWindowDelay for slave

transmitWindowDelay is either 2.5ms or 3.75ms when AUX_CONNECT_REQ is
used (for 1M/2M and Coded respectively). It also uses the same unit
as transmitWindowOffset (1.25ms) so we can just simply add both.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Use proper channel sel algorithm

Connections created from extended advertising instance always use CSA2.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Use proper phy when initializing slave conn

Connection created from extended advertising instance uses the same phy
as on secondary channels.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Initialize conn for connectable ext adv

Connection structs need to be initialized when advertising is enabled
on connectable extended advertising instance.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add support for connectable ext adv

This adds handling for received AUX_CONNECT_REQ PDU which enables slave
connection with extended advertising.

Once AUX_CONNECT_REQ is received, standard checks are performed to
determine if we have resource to create connection, however unlike for
CONNECT_IND an rx node is temporarily kept in LLL until AUX_CONNECT_RSP
is successfully sent. This is to prevent creating a connection in case
response PDU was not sent for whatever reason.

A separate scratch buffer is created for AUX_CONNECT_RSP since default
radio scratch buffer is already used by received AUX_CONNECT_REQ.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -477,9 +543,102 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+	lll = ftr->param;
+
+	if (!radio_is_done()) {
+		/* TODO: release node_rx somehow */
```
I am not sure how to do this properly: free it here somehow (did not see any code that does similar thing) or send and handle as error in ull?

Basically node_rx is allocated when AUX_CONNECT_REQ is received because we at that point in time we need to decide whether we have resources to accept connection and also that PDU is used as an anchor point. However, we should not send node until AUX_CONNECT_RSP is sent as otherwise we may end with a connection that was not properly established.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -433,6 +490,15 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+	hdr = (void *)&pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[0];
+	if (hdr->tgt_addr) {
+		/* TargetA is placed at 8th byte of ext hdr data */
+		tgt_addr = &pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[7];
```
Consider adding `#define`s for these indexes. 

Also consider putting 
```
hdr = (void *)&pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[0];
``` 
before 
```
addr = &pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[1];
```
for semantic reasons :)

andrzej-kaczmarek replied:
this will be addressed once https://github.com/zephyrproject-rtos/zephyr/pull/29895 is merged

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -477,9 +543,102 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+		radio_gpio_pa_setup();
+		radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() +
+					 EVENT_IFS_US -
+					 radio_rx_chain_delay_get(0, 0) -
```
Consider replacing `(0, 0)` with two local variables, `uint8_t phy = 0` and `uint8_t flags = 0` so it's clearer what the values are. Should be done in line 602 as well, which will also ensure that if those values every become different from 0, it only needs to be updated a single place.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -477,9 +543,102 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+		rx = ull_pdu_rx_alloc();
+
+		rx->hdr.type = NODE_RX_TYPE_CONNECTION;
+		rx->hdr.handle = 0xffff;
```
Not necessarily part of the scope of this PR, but I see that `0xFFFF` is used multiple places for an initial handle. I would strongly consider making a `#define` for that to be used going forward

Thalley reviewed (commented):
A few minor non-issue comments from me

andrzej-kaczmarek committed:
Bluetooth: controller: Fix phy initialization when enabling adv

Make sure phy is properly initialized to 1M when ext adv is supported
and legacy advertising instance is being enabled.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Make global helper to get aux handle

Just make local function a global one, we need this in other files
as well.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Stop aux after connected

After slave connection is created only adv instance is stopped, we also
need to stop aux separately if enabled.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -696,9 +729,16 @@ static inline int isr_rx_pdu(struct lll_adv *lll,
 	addr = pdu_adv->adv_ind.addr;
 	tx_addr = pdu_adv->tx_addr;
 
+	if (pdu_adv->type == PDU_ADV_TYPE_DIRECT_IND) {
```
correct, line 733 should be `tgt_addr = pdu_adv->direct_ind.tgt_addr`

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -429,14 +486,18 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
-		/* AdvA is placed at 2nd byte of ext hdr data */
-		addr = &pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[1];
-		tx_addr = pdu_aux->tx_addr;
+	/* AdvA is placed at 2nd byte of ext hdr data */
```
Yes, since this code handles aux PDUs it means we always have `ADV_EXT_IND` on primary, otherwise we have legacy advertising and no aux PDUs to handle. 

btw, this change was already merged as a part of some other PR iirc so will be removed from this PR

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -482,9 +543,102 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+		struct node_rx_pdu *rx;
+
+		if (IS_ENABLED(CONFIG_BT_CTLR_CHAN_SEL_2)) {
+			rx = ull_pdu_rx_alloc_peek(4);
```
tbh this was copied from `lll_adv.c` ;)
Anyway, we want either 1 or 2 buffers but there are apparently 2 reserved buffers - see comment in `lll_conn.c:769` - so need to check for 3 or 4 instead. Probably would be good to address this separately and have some `#define` or dedicated helper.

andrzej-kaczmarek committed:
Bluetooth: controller: Release resources on failed AUX_CONNECT_RSP

We allocate nodes on receiving AUX_CONNECT_REQ, but we can only use
them if AUX_CONNECT_RSP is sent successfully. If that fails, we need
to release those nodes.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Allow to simply release node allocated in LLL

If we allocate a node in LLL and it turns out that we don't need it
(e.g. allocated connection on AUX_CONNECT_REQ, but connection handshake
did not complete and we don't need it) we still need to send it to ULL
to be released. We can use existing NODE_RX_TYPE_DC_PDU_RELEASE node
type for this purpose, but we just need to make sure it's passed from
LLL to ULL properly.

Also, since this now does not only release DC PDUs, let's change its
name to a more generic one.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -469,9 +536,116 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+			return -ENOBUFS;
+		}
+
+		radio_isr_set(isr_tx_connect_rsp, rx);
```
Add a comment here that `rx` is allocated from the free rx queue later in this function, so that ISR critical steps is done first.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -469,9 +536,116 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+	lll = ftr->param;
+
+	if (!radio_is_done()) {
+		/* AUX_CONNECT_RSP was not sent properly, need to release
+		 * allocated resources.
+		 */
+
+		rx->hdr.type = NODE_RX_TYPE_RELEASE;
+		ull_rx_put(rx->hdr.link, rx);
+
+		if (IS_ENABLED(CONFIG_BT_CTLR_CHAN_SEL_2)) {
+			rx = ftr->extra;
+			rx->hdr.type = NODE_RX_TYPE_RELEASE;
+			ull_rx_put(rx->hdr.link, rx);
+		}
+
+		ull_rx_sched();
+		return;
```
```suggestion
		/* AUX_CONNECT_RSP was not sent properly, need to release
		 * allocated resources.
		 */

		rx->hdr.type = NODE_RX_TYPE_DC_PDU_RELEASE;

		if (IS_ENABLED(CONFIG_BT_CTLR_CHAN_SEL_2)) {
			ull_rx_put(rx->hdr.link, rx);

			rx = ftr->extra;
			rx->hdr.type = NODE_RX_TYPE_DC_PDU_RELEASE;
		}
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -469,9 +536,116 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
+		return;
+	}
+
+	/* Clear radio status and events */
+	lll_isr_status_reset();
+	lll_isr_cleanup(lll);
+
+	/* Stop further LLL radio events */
+	ret = lll_stop(lll);
+	LL_ASSERT(!ret);
+
+	ull_rx_put(rx->hdr.link, rx);
+	ull_rx_sched();
```
```suggestion
	ull_rx_put(rx->hdr.link, rx);
	ull_rx_sched();

	/* Clear radio status and events */
	lll_isr_status_reset();
	lll_isr_cleanup(lll);

	/* Stop further LLL radio events */
	ret = lll_stop(lll);
	LL_ASSERT(!ret);
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -99,6 +100,19 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
 	win_offset = sys_le16_to_cpu(pdu_adv->connect_ind.win_offset);
 	conn_interval_us = interval * 1250U;
 
+	if (0) {
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	} else if (adv->lll.aux) {
+		if (adv->lll.phy_s & BIT(2)) {
+			win_delay = 3;
+		} else {
+			win_delay = 2;
+		}
+#endif
+	} else {
+		win_delay = 1;
+	}
+
```
Is this 1, 2 and 3 BT Spec constants? define them in pdu.h or lll_tim_internal.h

cvinayak reviewed (changes requested):
Update tests/bluetooth/bsim_bt/bsim_test_advx will connectable extended advertising scenario.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -421,10 +478,20 @@ static inline int isr_rx_pdu(struct lll_adv_aux *lll_aux,
 	addr = &pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[1];
 	tx_addr = pdu_aux->tx_addr;
 
+	if (hdr->tgt_addr) {
+		/* TargetA is placed at 8th byte of ext hdr data */
+		tgt_addr = &pdu_aux->adv_ext_ind.ext_hdr_adi_adv_data[7];
+	} else {
+		tgt_addr = NULL;
+	}
+	rx_addr = pdu_aux->rx_addr;
+
```
`pdu_aux` is an AUX PDU we sent before RX, there's no need to check its type

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -157,6 +157,8 @@ struct lll_event {
 enum node_rx_type {
 	/* Unused */
 	NODE_RX_TYPE_NONE = 0x00,
+	/* Signals simple release of node */
+	NODE_RX_TYPE_RELEASE,
```
That type is used only in ULL, it's never sent from LLL. I can add add handling so it can be also sent from LLL, but perhaps we should also rename it to `NODE_RX_TYPE_RELEASE` since it will now release non-DC PDUs also?

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -39,6 +39,11 @@
 #include "common/log.h"
 #include "hal/debug.h"
 
+static uint8_t lll_adv_connect_rsp_pdu[PDU_AC_LL_HEADER_SIZE +
```
I can use scratch buffer, but since RX PDU is already there I need to allocate rx node and copy it there before it can be used for TX. This means critical ISR code is longer. Having this in separate buffer also allows to optimize by initializing PDU only once since we only need to update AdvA and TargetA for each connection.

andrzej-kaczmarek commented:
> Update tests/bluetooth/bsim_bt/bsim_test_advx will connectable extended advertising scenario.

I'll add this in a PR that adds support for connectable extended advertising in initiator role, otherwise those tests would not work.

andrzej-kaczmarek committed:
Bluetooth: controller: Call lll_adv_aux initialization funcs

lll_adv_aux_init and lll_adv_aux_reset were not called anywhere.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -117,6 +114,15 @@ int lll_adv_init(void)
 {
 	int err;
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_CTLR_ADV_AUX_SET)
```
Nitpick: `#if defined(CONFIG_BT_CTLR_ADV_EXT)` isn't really necessary here

cvinayak replied:
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_EXT)
#if (BT_CTLR_ADV_AUX_SET > 0)
```

note, without `CONFIG_`

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -357,6 +372,39 @@ int lll_adv_scan_req_report(struct lll_adv *lll, struct pdu_adv *pdu_adv_rx,
+	}
+
+#if defined(CONFIG_BT_CTLR_PRIVACY)
+	return ((((lll->filter_policy & 0x02) == 0) &&
```
Out of scope of this PR, but something I'd like to see in the future: `#define`d values for the filter_policy.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -790,19 +797,35 @@ uint8_t ll_adv_enable(uint8_t enable)
 		conn_lll->phy_flags = 0;
-		conn_lll->phy_tx_time = BIT(0);
-		conn_lll->phy_rx = BIT(0);
+		if (0) {
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+		} else if (pdu_adv->type == PDU_ADV_TYPE_EXT_IND) {
+			conn_lll->phy_tx = lll->phy_s;
+			conn_lll->phy_tx_time = lll->phy_s;
+			conn_lll->phy_rx = lll->phy_s;
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+		} else {
+			conn_lll->phy_tx = BIT(0);
+			conn_lll->phy_tx_time = BIT(0);
+			conn_lll->phy_rx = BIT(0);
+		}
```
Is the `if (IS_ENABLED(...)) { ... } ` not something we use in the controller? 

Otherwise, this code could have been
```
		if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) && 
		    pdu_adv->type == PDU_ADV_TYPE_EXT_IND)) {
			conn_lll->phy_tx = lll->phy_s;
			conn_lll->phy_tx_time = lll->phy_s;
			conn_lll->phy_rx = lll->phy_s;
		} else {
			conn_lll->phy_tx = BIT(0);
			conn_lll->phy_tx_time = BIT(0);
			conn_lll->phy_rx = BIT(0);
		}
```

andrzej-kaczmarek replied:
it is used, but in this case `phy_s` if conditionally included in structure so need to `#ifdef` here as well

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -99,6 +100,19 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
+	if (0) {
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	} else if (adv->lll.aux) {
+		if (adv->lll.phy_s & BIT(2)) {
```
Don't have some PHY macros that we could use instead of `BIT(2)`? I know there's `BT_HCI_LE_EXT_SCAN_PHY_CODED` but not sure if we want to use something with the `SCAN` infix here.

andrzej-kaczmarek replied:
We do, there are `PHY_xxx` macros but they are not quite used. I started separate change where it's updated everywhere in code so will address this in another PR.

Thalley reviewed (approved):
Only some minor nitpicking comments from me. Worth taking a look at them, and potentially fix some of them, but not necessary.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -129,6 +135,15 @@ int lll_adv_reset(void)
 	int err;
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_CTLR_ADV_AUX_SET)
```
```suggestion
#if (BT_CTLR_ADV_AUX_SET > 0)
```

note, without `CONFIG_`

andrzej-kaczmarek replied:
@cvinayak fixed

cvinayak reviewed (changes requested):
This can be handled in a separate PR: https://github.com/zephyrproject-rtos/zephyr/pull/28754/files#r521950803

Just two minor change requests to use `#if (BT_CTLR_ADV_AUX_SET > 0)` : 
https://github.com/zephyrproject-rtos/zephyr/pull/28754/files#r533704778

Otherwise, tested the connectable advertising and able to successfully establish connections (analyzed with sniffer logs and PDUs appear correct).

andrzej-kaczmarek commented:
rebased to resolve conflicts
